### PR TITLE
Integrate with argo workflow via vineyardctl.

### LINKED ISSUE #1590: CSI Driver: Integrate with argo workflow via vineyardctl
Describe your problem
---------------------

At present, we have integrated kubeflow through CSI Driver. However, kubeflow's resource scheduling is handled by argo workflow. Thus, we can integrate with argo workflow directly.

Expected Behavior
-------------------

1. Install the Vineyard CSI Driver while installing the Argo Server.
2. Change APIs like `pd.read_pickle/write_pickle` to `vineyard.csi.write/read` in the source code and then push images by users.
3. Implement a subcommand such as `vineyardctl inject argoworkflow --input user_argo_workflow.yaml --template-name="xxx" --vineyard-cluster="xxx" --path="xxx"`.

- User input an argo workflow yaml `argo_workflow.yaml` as follows
```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: mlops-
spec:
  entrypoint: dag
  templates:
  - name: MLops
    inputs:
      parameters:
      - name: functions
    container:
      imagePullPolicy: IfNotPresent
      image: mlops-benchmark:latest
      command: [python]
      args: ["-m",  "{{inputs.parameters.functions}}"]
  - name: dag
    dag:
      tasks:
      - name: prepare-data
        template: MLops
        arguments:
          parameters:
          - name: functions
            value: prepare-data.py
      - name: preprocess-data
        template: MLops
        dependencies:
          - prepare-data
        arguments:
          parameters:
          - name: functions
            value: preprocess-data.py
      - name: train-data
        template: MLops
        dependencies:
          - preprocess-data
        arguments:
          parameters:
          - name: functions
            value: train-data.py
      - name: test-data
        template: MLops
        dependencies:
          - train-data
        arguments:
          parameters:
          - name: functions
            value: test-data.py
```

- Run `vineyardctl inject argoworkflow argo_workflow.yaml --template-name=MLops --vineyard-cluster="vineyard-system/vineyardd-sample" --path="/vineyard/data" --dag-task="drop_cols_node, fill-na-node"` and get the following result.

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: mlops-
spec:
  entrypoint: dag
  templates:
  - name: MLops
    inputs:
      parameters:
      - name: functions
      - name: vineyard-objects-name
    container:
      imagePullPolicy: IfNotPresent
      image: mlops-benchmark:latest
      command: [python]
      args: ["-m",  "{{inputs.parameters.functions}}"]
      ########################added########################
      volumeMounts:
      - {mountPath: /vineyard/data, name: vineyard-objects}
      #####################################################
    #####################################added#########################################
    volumes:
    - name: vineyard-objects
      persistentVolumeClaim: {claimName: '{{inputs.parameters.vineyard-objects-name}}'}
    ###################################################################################
  #### add vineyard-objects as a dependency for each dag task using the specific template ####
  - name: dag
    dag:
      tasks:
      - name: prepare-data
        template: MLops
        dependencies:
          ############added###############
          - vineyard-objects
          ################################
        arguments:
          parameters:
          - name: functions
            value: prepare-data
          ######################################added###################################
          - name: vineyard-objects-name
            value: '{{tasks.vineyard-objects.outputs.parameters.vineyard-objects-name}}'
          ##############################################################################
      - name: preprocess-data
        template: MLops
        dependencies:
          - prepare-data
          ############added###############
          - vineyard-objects
          ################################
        arguments:
          parameters:
          - name: functions
            value: preprocess-data
          ######################################added###################################
          - name: vineyard-objects-name
            value: '{{tasks.vineyard-objects.outputs.parameters.vineyard-objects-name}}'
          ##############################################################################
      - name: train-data
        template: MLops
        dependencies:
          - preprocess-data
          ############added###############
          - vineyard-objects
          ################################
        arguments:
          parameters:
          - name: functions
            value: preprocess-data
          ######################################added###################################
          - name: vineyard-objects-name
            value: '{{tasks.vineyard-objects.outputs.parameters.vineyard-objects-name}}'
          ##############################################################################
      - name: test-data
        template: MLops
        dependencies:
          - train-data
          ############added###############
          - vineyard-objects
          ################################
        arguments:
          parameters:
          - name: functions
            value: test-data
          ######################################added###################################
          - name: vineyard-objects-name
            value: '{{tasks.vineyard-objects.outputs.parameters.vineyard-objects-name}}'
          ##############################################################################
      #########################added########################
      - {name: vineyard-objects, template: vineyard-objects}
      ######################################################
  ############################added###########################
  - name: vineyard-objects
    resource:
      action: create
      setOwnerReference: true
      manifest: |
        apiVersion: v1
        kind: PersistentVolumeClaim
        metadata:
          name: '{{workflow.name}}-vineyard-objects-pvc'
        spec:
          accessModes:
          - ReadWriteMany
          resources:
            requests:
              storage: 1Mi
          storageClassName: vineyard-system.vineyardd-sample.csi
    outputs:
      parameters:
      - name: vineyard-objects-name
        valueFrom: {jsonPath: '{.metadata.name}'}
  ##############################################################
```

/cc @sighingnow, @septicmk, @vegetableysm

What do you think? Any comment is welcome! Thanks.
 

There may be no user directly using the argo workflow to run a machine learning model, but instead the productized workflow engine, such as kubeflow or kerdo. That means the feature will not bring any gain for argo workflow user, as they don't need to pass dataframes between DAG tasks. However, most productized  ML workflow engines, such as [metaflow](https://docs.metaflow.org/production/scheduling-metaflow-flows/scheduling-with-argo-workflows), use the argo workflow as the scheduling backend, this feature may provide a chance to accelerate the dataframe sharing. In this way, we don't need to integrate with various ML workflow engine again and again.

Additional Context
-------------------

https://v6d.io/tutorials/kubernetes/efficient-data-sharing-in-kubeflow-with-vineyard-csi-driver.html#efficient-data-sharing-in-kubeflow-with-vineyard-csi-driver

### PR DESCRIPTION
**Pull request by dashanji**
What do these changes do?
-------------------------

* Add a new command 'vineyardctl inject argo-workflow' to inject the vineyard volumes to the workflow manifest.
* Add the relevant unit test.
* Update the doc.

Related issue number
--------------------

Fixes #1590

### CODE REVIEW

File: k8s/cmd/commands/inject/inject_argo_workflow.go [hunk 1]
```diff
@@ -0,0 +1,666 @@
...
+	
+	# Suppose you only have a single template in the workflow
+	# you could set only one template name in the --templates flag
+	$ vineyardctl inject argo-workflow -f workflow.yaml \
+         --templates="mlops" \
+		 --vineyard-cluster="vineyard-system/vineyardd-sample" \
+		 --mount-path="/vineyard/data" \
+		 --dag="dag" \
+		 --tasks="preprocess-data,test-data"
+	`)
```
Comment (sighingnow) [inline k8s/cmd/commands/inject/inject_argo_workflow.go:399]: Can you use a simpler demo case in docstring? for example a workflow with only 2 step.

File: k8s/cmd/commands/inject/inject_argo_workflow_test.go [hunk 2]
```diff
@@ -0,0 +1,388 @@
...
+package inject
+
+import (
+	"os"
+	"reflect"
+	"testing"
+
+	"github.com/v6d-io/v6d/k8s/cmd/commands/flags"
+	"github.com/v6d-io/v6d/k8s/cmd/commands/util"
+	"gopkg.in/yaml.v3"
```
Comment (sighingnow) [inline k8s/cmd/commands/inject/inject_argo_workflow_test.go:25]: Leave vineyard imports to the last position.

**Review by sighingnow — APPROVED** [review]

File: k8s/cmd/commands/inject/inject_argo_workflow.go (hunk 1 above)
Comment (dashanji) [reply k8s/cmd/commands/inject/inject_argo_workflow.go:399]: Ok.

File: k8s/cmd/commands/inject/inject_argo_workflow_test.go (hunk 2 above)
Comment (dashanji) [reply k8s/cmd/commands/inject/inject_argo_workflow_test.go]: Done.

**Pull request merged**